discovery/syncer.go

```diff
@@ -0,0 +1,895 @@
+
+	// bufferedChanRangeReplies is used in the waitingQueryChanReply to
+	// buffer all the chunked response to our query.
+	bufferedChanRangeReplies []lnwire.ShortChannelID
+
+	// newChansToQuery is used to pass the set of channels we should query
+	// for from the waitingQueryChanReply state to the queryNewChannels
+	// state.
+	newChansToQuery []lnwire.ShortChannelID
+
+	// peerPub is the public key of the peer we're syncing with, serialized
+	// in compressed format.
+	peerPub [33]byte
+
+	cfg gossipSyncerCfg
+
+	sync.Mutex
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// newGossiperSyncer returns a new instance of the gossipSyncer populated using
+// the passed config.
+func newGossiperSyncer(cfg gossipSyncerCfg) *gossipSyncer {
+	return &gossipSyncer{
+		cfg:        cfg,
+		gossipMsgs: make(chan lnwire.Message, 100),
+		quit:       make(chan struct{}),
+	}
+}
+
+// Start starts the gossipSyncer and any goroutines that it needs to carry out
+// it duties.
+func (g *gossipSyncer) Start() error {
+	log.Debugf("Starting gossipSyncer(%x)", g.peerPub[:])
+
+	g.wg.Add(1)
+
+	go g.channelGraphSyncer()
+
+	return nil
+}
+
+// Stop signals the gossipSyncer for a graceful exit, then waits until it has
+// exited.
+func (g *gossipSyncer) Stop() error {
+	close(g.quit)
+
+	g.wg.Wait()
+
+	return nil
+}
+
+// channelGraphSyncer is the main goroutine responsible for ensuring that we
+// properly channel graph state with the remote peer, and also that we only
+// send them messages which actually pass their defined update horizon.
+func (g *gossipSyncer) channelGraphSyncer() {
+	defer g.wg.Done()
+
+	// TODO(roasbeef): also add ability to force transition back to syncing
+	// chans
+	//  * needed if we want to sync chan state very few blocks?
+
+	for {
+		log.Debugf("gossipSyncer(%x): state=%v", g.peerPub[:], g.state)
+
+		switch g.state {
+		// When we're in this state, we're trying to synchronize our
+		// view of the network with the remote peer. We'll kick off
+		// this sync by asking them for the set of channels they
+		// understand, as we'll as responding to any other queries by
+		// them.
+		case syncingChans:
+			// If we're in this state, then we'll send the remote
+			// peer our opening QueryChannelRange message.
+			queryRangeMsg, err := g.genChanRangeQuery()
+			if err != nil {
+				log.Errorf("unable to gen chan range "+
+					"query: %v", err)
+				return
+			}
+
+			err = g.cfg.sendToPeer(queryRangeMsg)
+			if err != nil {
+				log.Errorf("unable to send chan range "+
+					"query: %v", err)
+				return
+			}
+
+			// With the message sent successfully, we'll transition
+			// into the next state where we wait for their reply.
+			g.state = waitingQueryRangeReply
+
+		// In this state, we've sent out our initial channel range
+		// query and are waiting for the final response from the remote
+		// peer before we perform a diff to see with channels they know
+		// of that we don't.
+		case waitingQueryRangeReply:
```

Comment:
Could you use helper functions to reduce the amount of code that is in these case statements? Basically I'm imagining that for each top-level case, there would be one helper function that would contain almost all of the code this is currently contained in the case.

This would
- make channelGraphSyncer considerably shorter (easier to fit into one's head)
- create a clear separation between high-level details and low-level details (which would be contained in the helper functions)
- make unit testing way easier, cleaner, and more granular
- decrease vertical line noise by decreasing line-wrapping for over-indentation

Note that the helper functions will probably need multiple return values to allow for the "return", "continue", and standard case endings.

Reply:
All of  "transition" logic is already located in distinct functions. The logic in each of the case statements simply calls out to those, and handles the redundant case in each state. As is, the entire file already have a comprehensive set of unit tests for both the transition functions, and also the control flow interaction between two syncer instances. 

Generally, in the codebase, we favor a clear control flow over excessive modularization. Vertical space isn't as much of an issue following this philosophy.